### PR TITLE
Code cleanup and better Kubernetes usability

### PR DESCRIPTION
**Pull request by korkin25**
Hi, guys!

first of all: thank you for fine quality geoserver deploy.
I just fixed a few strange things and added ability to see geoserver nodes as Pod names.

Thank you!

### CODE REVIEW

**Comment by NyakudyaA** [comment]
@korkin25 are you done with this PR. Can I review and test it now

**Comment by korkin25** [comment]
Sure! thank you! IF you will find any bug, say me - I will fix:) 
I  use it in production right now. I have some issues (1. guava cache doesnt work 2. s3 caching doesnt work), but I not sure it's related to this repo directly.

File: scripts/entrypoint.sh [hunk 1]
```diff
@@ -7,16 +7,14 @@ figlet -t "Kartoza Docker GeoServer"
 source /scripts/functions.sh
 source /scripts/env-data.sh
 
-/bin/bash /scripts/start.sh
+set_vars
+export  READONLY CLUSTER_DURABILITY BROKER_URL EMBEDDED_BROKER TOGGLE_MASTER TOGGLE_SLAVE BROKER_URL
+export CLUSTER_CONFIG_DIR MONITOR_AUDIT_PATH CLUSTER_LOCKFILE INSTANCE_STRING
```
Comment (NyakudyaA) [inline scripts/entrypoint.sh:12]: The instance string is always going to be empty like the logs are showing
empty

File: scripts/functions.sh [hunk 2]
```diff
@@ -3,6 +3,10 @@
 
 export request="wget --progress=bar:force:noscroll -c --tries=2 "
 
+function log() {
```
Comment (NyakudyaA) [inline scripts/functions.sh:6]: What does this function do ?

File: scripts/functions.sh [hunk 3]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
+  generate_random_string 20 # Temporary not sure that that is need any more...
 
-  generate_random_string 20
+  
   if [[ -z ${INSTANCE_STRING} ]];then
-    INSTANCE_STRING=${RAND}
-  else
-    INSTANCE_STRING=${INSTANCE_STRING}
+    if [ ! -z "${HOSTNAME}" ]; then
+      INSTANCE_STRING="${HOSTNAME: -20}"    # Use hostname as INSTANCE_STRING. It's still compatible w/ docker-compose and Kubernetes (instance will be same as Pod names)
```
Comment (NyakudyaA) [inline scripts/functions.sh:311]: Where is the $HOSTNAME variable defined? I do not see any default value for it. What about older images how will they handle migration of data directories with clustering enabled?

File: scripts/start.sh [hunk 4]
```diff
@@ -34,6 +34,11 @@ create_dir "${GEOSERVER_DATA_DIR}"/logs
 export GEOSERVER_LOG_LEVEL
 geoserver_logging
 
+#create /gwc in case of missong that
+if [ ! -d "${GEOWEBCACHE_CACHE_DIR}" ]; then
+        create_dir "${GEOWEBCACHE_CACHE_DIR}"
```
Comment (NyakudyaA) [inline scripts/start.sh:39]: Isn't this duplication from https://github.com/kartoza/docker-geoserver/blob/master/Dockerfile#L48

File: README.md [hunk 5]
```diff
@@ -79,6 +79,7 @@ To build yourself with a local checkout using the docker-compose.build.yaml:
    DOWNLOAD_ALL_COMMUNITY_EXTENSIONS=Specifies whether to download all community plugins or a single one
    GEOSERVER_UID=Specifies the uid to use for the user used to run GeoServer in the container
    GEOSERVER_GID=Specifies the gid to use for the group used to run GeoServer in the container
+   GEOSERVER_PUBLISH_DIR=Specifies directory under ${CATALINA_HOME}/webapps where geoserver tomcat files will be installed
```
Comment (NyakudyaA) [inline README.md:82]: What does this variable do? Is there any time where users can specify a non-default location of tomcat? If so the script needs to handle how entrypoint will correspond to this.

**Comment by NyakudyaA** [comment]
> Sure! thank you! IF you will find any bug, say me - I will fix:) I use it in production right now. I have some issues (1. guava cache doesnt work 2. s3 caching doesnt work), but I am not sure it's related to this repo directly.

@korkin25 Can you address the issues I raised above

File: scripts/functions.sh (hunk 3 above)
Comment (korkin25) [reply scripts/functions.sh:311]: $HOSTNAME is always present in any pod/container. you can also always use $(hostname).

**Comment by korkin25** [comment]
About: function log() {

this is debug finction which provide better logging with line&filename&and function name if present.
for example:
/scripts/entrypoint.sh:33 87 0: Replacing S3 servers configuration from configmap...

About GEOSERVER_PUBLISH_DIR:

forget about that. In next commit I removed any my modification regarding of GEOSERVER_PUBLISH_DIR   (https://github.com/kartoza/docker-geoserver/pull/371/commits/66fea9336fd511a3d49f09785a9617671ea6eab9)

**Comment by NyakudyaA** [comment]
@korkin25 Can you resolve the conflicts by pulling from master whilst you at it

File: scripts/functions.sh [hunk 6]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
-    CLUSTER_CONFIG_DIR="${GEOSERVER_DATA_DIR}/cluster/instance_${RAND}"
-    MONITOR_AUDIT_PATH="${GEOSERVER_DATA_DIR}/monitoring/monitor_${RAND}"
-    CLUSTER_LOCKFILE="${CLUSTER_CONFIG_DIR}/.cluster.lock"
-  else
-
-    CLUSTER_CONFIG_DIR="${GEOSERVER_DATA_DIR}/cluster/instance_${RANDOMSTRING}"
-    MONITOR_AUDIT_PATH="${GEOSERVER_DATA_DIR}/monitoring/monitor_${RANDOMSTRING}"
-    CLUSTER_LOCKFILE="${CLUSTER_CONFIG_DIR}/.cluster.lock"
-  fi
+  generate_random_string 20 # Temporary not sure that that is need any more...
```
Comment (NyakudyaA) [inline scripts/functions.sh:306]: You can remove it. Not needed here

File: scripts/functions.sh (hunk 3 above)
Comment (NyakudyaA) [inline scripts/functions.sh:311]: change INSTANCE_STRING="${HOSTNAME: -20}" to `INSTANCE_STRING="${HOSTNAME: -12}"`. In docker-compose it has a 12 character limit.

File: scripts/functions.sh [hunk 7]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
-  else
-    INSTANCE_STRING=${INSTANCE_STRING}
+    if [ ! -z "${HOSTNAME}" ]; then
+      INSTANCE_STRING="${HOSTNAME: -20}"    # Use hostname as INSTANCE_STRING. It's still compatible w/ docker-compose and Kubernetes (instance will be same as Pod names)
+    else
+      INSTANCE_STRING=${RAND}
+    fi
   fi
 
+  generate_random_string 14 # Temporary not sure that that is need any more...
```
Comment (NyakudyaA) [inline scripts/functions.sh:317]: Remove not needed here

File: scripts/functions.sh [hunk 8]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
+
```
Comment (NyakudyaA) [inline scripts/functions.sh:318]: what happens here if I supply the env variable `INSTANCE_STRING` with less than 14 characters. Will it run?

File: scripts/functions.sh [hunk 9]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
+  CLUSTER_CONFIG_DIR="${GEOSERVER_DATA_DIR}/cluster/instance_${INSTANCE_STRING: -14}"
```
Comment (NyakudyaA) [inline scripts/functions.sh:319]: change -14 to 14 for consistency

File: scripts/functions.sh [hunk 10]
```diff
@@ -299,25 +303,22 @@ function file_env {
...
+  MONITOR_AUDIT_PATH="${GEOSERVER_DATA_DIR}/monitoring/monitor_${INSTANCE_STRING: -14}"
```
Comment (NyakudyaA) [inline scripts/functions.sh:320]: change -14 to 12 for consistency

**Comment by NyakudyaA** [comment]
Ping @korkin25

File: scripts/functions.sh [hunk 11]
```diff
@@ -3,6 +3,10 @@
 
 export request="wget --progress=bar:force:noscroll -c --tries=2 "
 
+function log() {
+    echo "$0:${BASH_LINENO[*]}": $@
```
Comment (dunn) [inline scripts/functions.sh:7]: The column number will always be `0`, so we only need the line number from the `BASH_LINENO` array, I think?

Comment (dunn) [reply scripts/functions.sh:7]: Currently:
```
shoreline-geoserver-1   | /scripts/entrypoint.sh:16 0: CLUSTER_CONFIG_DIR=/data/cluster/instance_
shoreline-geoserver-1   | /scripts/entrypoint.sh:17 0: MONITOR_AUDIT_PATH=/data/monitoring/monitor_
```

**Comment by NyakudyaA** [comment]
Superseeded by #401

**Pull request closed without merging**